File: packages/hquant/src/common/TypedRingBuffer.ts
// 支持 TypedArray 的高性能循环队列

export class TypedRingBuffer {
  private buffer: Float64Array | Int32Array;
  private capacity: number;
  private front: number = 0;
  private rear: number = 0;
  private _length: number = 0;

  constructor(type: 'float' | 'int', capacity: number) {
    this.capacity = capacity;
    this.buffer = type === 'float' ? new Float64Array(capacity) : new Int32Array(capacity);
  }

  push(value: number): boolean {
    if (this._length === this.capacity) {
      // 队列满，覆盖最旧数据
      this.front = (this.front + 1) % this.capacity;
      this._length--;
    }
    this.buffer[this.rear] = value;
    this.rear = (this.rear + 1) % this.capacity;
    this._length++;
    return true;
  }

  shift(): number | undefined {
    if (this._length === 0) return undefined;
    const value = this.buffer[this.front];
    this.front = (this.front + 1) % this.capacity;
    this._length--;
    return value;
  }

  pop(): number | undefined {
    if (this._length === 0) return undefined;
    this.rear = (this.rear - 1 + this.capacity) % this.capacity;
    const value = this.buffer[this.rear];
    this._length--;
    return value;
  }

  update(index: number, value: number): boolean {
    if (index < 0 || index >= this._length) return false;
    const i = (this.front + index) % this.capacity;
    this.buffer[i] = value;
    return true;
  }

  get(index: number): number | undefined {
    if (index < 0 || index >= this._length) return undefined;
    const i = (this.front + index) % this.capacity;
    return this.buffer[i];
  }

  getLast(): number | undefined {
    return this.get(this._length - 1);
  }

  size(): number {
    return this._length;
  }

  clear(): void {
    if (this.buffer instanceof Float64Array) {
      this.buffer = new Float64Array(this.capacity);
    } else {
      this.buffer = new Int32Array(this.capacity);
    }
    this.front = 0;
    this.rear = 0;
    this._length = 0;
  }

  toArray(): number[] {
    const arr: number[] = [];
    for (let i = 0; i < this._length; i++) {
      arr.push(this.get(i)!);
    }
    return arr;
  }
}

File: packages/hquant-go-v2/common/typed_ring_buffer.go

package common

import "sync"

// TypedRingBuffer is a circular buffer for float64 values.
type TypedRingBuffer struct {
	data    []float64
	maxSize int
	front   int
	rear    int
	length  int
	mutex   sync.RWMutex
}

// NewTypedRingBuffer creates a new TypedRingBuffer.
func NewTypedRingBuffer(maxSize int) *TypedRingBuffer {
	return &TypedRingBuffer{
		data:    make([]float64, maxSize),
		maxSize: maxSize,
	}
}

// Push adds a value to the buffer.
func (t *TypedRingBuffer) Push(value float64) {
	t.mutex.Lock()
	defer t.mutex.Unlock()

	if t.length == t.maxSize {
		t.front = (t.front + 1) % t.maxSize
		t.length--
	}
	t.data[t.rear] = value
	t.rear = (t.rear + 1) % t.maxSize
	t.length++
}

// Get returns a value at a specific index.
func (t *TypedRingBuffer) Get(index int) float64 {
	t.mutex.RLock()
	defer t.mutex.RUnlock()

	if index < 0 {
		index = t.length + index
	}

	if index < 0 || index >= t.length {
		return 0 // Or handle error appropriately
	}

	i := (t.front + index) % t.maxSize
	return t.data[i]
}

// Update updates a value at a specific index.
func (t *TypedRingBuffer) Update(index int, value float64) {
	t.mutex.Lock()
	defer t.mutex.Unlock()

	if index < 0 {
		index = t.length + index
	}

	if index < 0 || index >= t.length {
		return // Or handle error appropriately
	}

	i := (t.front + index) % t.maxSize
	t.data[i] = value
}

// Size returns the number of items in the buffer.
func (t *TypedRingBuffer) Size() int {
	t.mutex.RLock()
	defer t.mutex.RUnlock()
	return t.length
}

func (t *TypedRingBuffer) Clear() {
	t.mutex.Lock()
	defer t.mutex.Unlock()
	t.data = make([]float64, t.maxSize)
	t.front = 0
	t.rear = 0
	t.length = 0
}

// ToArray returns the buffer as a slice.
func (t *TypedRingBuffer) ToArray() []float64 {
	t.mutex.RLock()
	defer t.mutex.RUnlock()

	result := make([]float64, t.length)
	for i := 0; i < t.length; i++ {
		result[i] = t.Get(i)
	}
	return result
}
